tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8568,6 +8703,105 @@ void hlsl::DiagnoseRegisterType(
   }
 }
 
+struct NameLookup {
+  FunctionDecl *Found;
+  FunctionDecl *Other;
+};
+
+static NameLookup GetSingleFunctionDeclByName(clang::Sema *self, StringRef Name, bool checkPatch) {
+  auto DN = DeclarationName(&self->getASTContext().Idents.get(Name));
+  FunctionDecl *pFoundDecl = nullptr;
+  for (auto idIter = self->IdResolver.begin(DN), idEnd = self->IdResolver.end(); idIter != idEnd; ++idIter) {
+    FunctionDecl *pFnDecl = dyn_cast<FunctionDecl>(*idIter);
+    if (!pFnDecl) continue;
+    if (checkPatch && !self->getASTContext().IsPatchConstantFunctionDecl(pFnDecl)) continue;
+    if (pFoundDecl) {
+      return NameLookup{ pFoundDecl, pFnDecl };
+    }
+    pFoundDecl = pFnDecl;
+  }
+  return NameLookup{ pFoundDecl, nullptr };
+}
+
+void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
+  DXASSERT_NOMSG(self != nullptr);
+
+  // Don't bother with global validation if compilation has already failed.
+  if (self->getDiagnostics().hasErrorOccurred()) {
+    return;
+  }
+
+  // TODO: make these error 'real' errors rather than on-the-fly things
+  // Validate that the entry point is available.
+  ASTContext &Ctx = self->getASTContext();
+  DiagnosticsEngine &Diags = self->getDiagnostics();
+  FunctionDecl *pEntryPointDecl = nullptr;
+  FunctionDecl *pPatchFnDecl = nullptr;
+  const std::string &EntryPointName = self->getLangOpts().HLSLEntryFunction;
+  if (!EntryPointName.empty()) {
+    NameLookup NL = GetSingleFunctionDeclByName(self, EntryPointName, /*checkPatch*/ false);
+    if (NL.Found && NL.Other) {
+      // NOTE: currently we cannot hit this codepath when CodeGen is enabled, because
+      // CodeGenModule::getMangledName will mangle the entry point name into the bare
+      // string, and so ambiguous points will produce an error earlier on.
+      unsigned id = Diags.getCustomDiagID(clang::DiagnosticsEngine::Level::Error,
+        "ambiguous entry point function");
+      Diags.Report(NL.Found->getSourceRange().getBegin(), id);
+      Diags.Report(NL.Other->getLocation(), diag::note_previous_definition);
+      return;
+    }
+    pEntryPointDecl = NL.Found;
+    if (!pEntryPointDecl || !pEntryPointDecl->hasBody()) {
+      unsigned id = Diags.getCustomDiagID(clang::DiagnosticsEngine::Level::Error,
+        "missing entry point definition");
+      Diags.Report(id);
+      return;
+    }
+  }
+
+  // Validate that there is no recursion; start with the entry function.
+  // NOTE: the information gathered here could be used to bypass code generation
+  // on functions that are unreachable (as an early form of dead code elimination).
+  if (pEntryPointDecl) {
```

Comment:
I don't think so - there's a different error that will kick in if we looked for the entry point but couldn't find it and we wouldn't reach here. We would hit this codepath if the entry point isn't specified at all, in which case we're probably running in an IDE, and then we shouldn't be running this bit of analysis.